File: pulumi/library/go/pkg/cicd/gitlab.go
/*
 * Copyright 2026 Vitruvian Software
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cicd

import (
	"fmt"

	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/iam"
	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/serviceaccount"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// GitLabOIDCArgs configures a GitLab OIDC provider, matching the
// 0-bootstrap/modules/gitlab-oidc local module in the Terraform foundation.
type GitLabOIDCArgs struct {
	// ProjectID is the GCP project in which to create the WIF resources.
	ProjectID pulumi.StringInput
	// PoolID is the Workload Identity Pool ID (e.g., "foundation-pool").
	PoolID pulumi.StringInput
	// ProviderID is the WIF Provider ID (e.g., "foundation-gl-provider").
	ProviderID pulumi.StringInput
	// AttributeCondition is the CEL expression restricting which tokens are accepted.
	// Example: "assertion.project_path.startsWith('my-group/')"
	AttributeCondition pulumi.StringInput
	// SAMapping maps logical keys to service account + attribute binding pairs.
	SAMapping map[string]SAMappingEntry
	// IssuerUri is the GitLab OIDC issuer URL. Defaults to "https://gitlab.com".
	IssuerUri pulumi.StringInput
}

// GitLabOIDC represents a GitLab Workload Identity Federation configuration.
type GitLabOIDC struct {
	pulumi.ResourceState
	Pool     *iam.WorkloadIdentityPool
	Provider *iam.WorkloadIdentityPoolProvider
	Bindings []*serviceaccount.IAMMember
}

// NewGitLabOIDC creates a new GitLabOIDC component resource.
func NewGitLabOIDC(ctx *pulumi.Context, name string, args *GitLabOIDCArgs, opts ...pulumi.ResourceOption) (*GitLabOIDC, error) {
	if args == nil {
		return nil, fmt.Errorf("args cannot be nil")
	}
	component := &GitLabOIDC{}
	err := ctx.RegisterComponentResource("pkg:cicd:GitLabOIDC", name, component, opts...)
	if err != nil {
		return nil, err
	}

	pool, err := iam.NewWorkloadIdentityPool(ctx, name+"-pool", &iam.WorkloadIdentityPoolArgs{
		Project:                args.ProjectID,
		WorkloadIdentityPoolId: args.PoolID,
		Disabled:               pulumi.Bool(false),
	}, pulumi.Parent(component))
	if err != nil {
		return nil, err
	}
	component.Pool = pool

	issuerUri := args.IssuerUri
	if issuerUri == nil {
		issuerUri = pulumi.String("https://gitlab.com")
	}

	// Attribute mapping matches the full default set from the TF gitlab-oidc
	// module variables.tf (13 attributes), ensuring parity with the upstream.
	provider, err := iam.NewWorkloadIdentityPoolProvider(ctx, name+"-provider", &iam.WorkloadIdentityPoolProviderArgs{
		Project:                        args.ProjectID,
		WorkloadIdentityPoolId:         pool.WorkloadIdentityPoolId,
		WorkloadIdentityPoolProviderId: args.ProviderID,
		AttributeCondition:             args.AttributeCondition,
		AttributeMapping: pulumi.StringMap{
			// Principal IAM
			"google.subject": pulumi.String("assertion.sub"),
			// Standard OIDC claims
			"attribute.sub": pulumi.String("assertion.sub"),
			"attribute.iss": pulumi.String("assertion.iss"),
			"attribute.aud": pulumi.String("assertion.aud"),
			"attribute.exp": pulumi.String("assertion.exp"),
			"attribute.nbf": pulumi.String("assertion.nbf"),
			"attribute.iat": pulumi.String("assertion.iat"),
			"attribute.jti": pulumi.String("assertion.jti"),
			// GitLab custom claims
			"attribute.namespace_id":   pulumi.String("assertion.namespace_id"),
			"attribute.namespace_path": pulumi.String("assertion.namespace_path"),
			"attribute.project_id":     pulumi.String("assertion.project_id"),
			"attribute.project_path":   pulumi.String("assertion.project_path"),
			"attribute.user_id":        pulumi.String("assertion.user_id"),
			"attribute.user_login":     pulumi.String("assertion.user_login"),
			"attribute.user_email":     pulumi.String("assertion.user_email"),
		},
		Oidc: &iam.WorkloadIdentityPoolProviderOidcArgs{
			IssuerUri: issuerUri,
		},
	}, pulumi.Parent(pool))
	if err != nil {
		return nil, err
	}
	component.Provider = provider

	for key, entry := range args.SAMapping {
		member := pulumi.Sprintf("principalSet://iam.googleapis.com/%s/%s", pool.Name, entry.Attribute)
		binding, err := serviceaccount.NewIAMMember(ctx, fmt.Sprintf("%s-binding-%s", name, key), &serviceaccount.IAMMemberArgs{
			ServiceAccountId: entry.SAName,
			Role:             pulumi.String("roles/iam.workloadIdentityUser"),
			Member:           member,
		}, pulumi.Parent(provider))
		if err != nil {
			return nil, err
		}
		component.Bindings = append(component.Bindings, binding)
	}

	ctx.RegisterResourceOutputs(component, pulumi.Map{
		"poolName":     pool.Name,
		"providerName": provider.Name,
	})

	return component, nil
}

File: packages/pulumi/library/ts/packages/cicd/src/gitlab.ts
import * as pulumi from "@pulumi/pulumi";
import * as gcp from "@pulumi/gcp";

export interface SAMappingEntry {
  saName: pulumi.Input<string>;
  attribute: string;
}

export interface GitLabOIDCArgs {
  projectId: pulumi.Input<string>;
  poolId: pulumi.Input<string>;
  providerId: pulumi.Input<string>;
  attributeCondition: pulumi.Input<string>;
  saMapping: Record<string, SAMappingEntry>;
  issuerUri?: pulumi.Input<string>;
}

export class GitLabOIDC extends pulumi.ComponentResource {
  public readonly pool: gcp.iam.WorkloadIdentityPool;
  public readonly provider: gcp.iam.WorkloadIdentityPoolProvider;
  public readonly bindings: gcp.serviceaccount.IAMMember[] = [];

  constructor(
    name: string,
    args: GitLabOIDCArgs,
    opts?: pulumi.ComponentResourceOptions,
  ) {
    super("pkg:cicd:GitLabOIDC", name, args, opts);

    this.pool = new gcp.iam.WorkloadIdentityPool(
      `${name}-pool`,
      {
        project: args.projectId,
        workloadIdentityPoolId: args.poolId,
        disabled: false,
      },
      { parent: this },
    );

    this.provider = new gcp.iam.WorkloadIdentityPoolProvider(
      `${name}-provider`,
      {
        project: args.projectId,
        workloadIdentityPoolId: this.pool.workloadIdentityPoolId,
        workloadIdentityPoolProviderId: args.providerId,
        attributeCondition: args.attributeCondition,
        attributeMapping: {
          "google.subject": "assertion.sub",
          "attribute.sub": "assertion.sub",
          "attribute.iss": "assertion.iss",
          "attribute.aud": "assertion.aud",
          "attribute.exp": "assertion.exp",
          "attribute.nbf": "assertion.nbf",
          "attribute.iat": "assertion.iat",
          "attribute.jti": "assertion.jti",
          "attribute.namespace_id": "assertion.namespace_id",
          "attribute.namespace_path": "assertion.namespace_path",
          "attribute.project_id": "assertion.project_id",
          "attribute.project_path": "assertion.project_path",
          "attribute.user_id": "assertion.user_id",
          "attribute.user_login": "assertion.user_login",
          "attribute.user_email": "assertion.user_email",
        },
        oidc: {
          issuerUri: args.issuerUri ?? "https://gitlab.com",
        },
      },
      { parent: this.pool },
    );

    for (const [key, entry] of Object.entries(args.saMapping)) {
      const member = pulumi.interpolate`principalSet://iam.googleapis.com/${this.pool.name}/${entry.attribute}`;
      this.bindings.push(
        new gcp.serviceaccount.IAMMember(
          `${name}-binding-${key}`,
          {
            serviceAccountId: entry.saName,
            role: "roles/iam.workloadIdentityUser",
            member: member,
          },
          { parent: this.provider },
        ),
      );
    }

    this.registerOutputs({
      poolName: this.pool.name,
      providerName: this.provider.name,
    });
  }
}
